Jenkinsfile

```diff
@@ -10,25 +34,267 @@ def ircNotification(stage, status) {
     sh "bin/irc-notify.sh --stage '${stage}' --status '${status}'"
 }
 
-node {
-    def branch = env.BRANCH_NAME
+def buildDockerImage(Map kwargs) {
+    def update = kwargs.update ? 'true' : 'false'
+    def repo = kwargs.dockerRepo ?: 'mozorg'
+    def script = kwargs.script ?: 'build_image.sh'
+    def environs = ["UPDATE_DOCKER_IMAGES=${update}",
+                    "DOCKERFILE=${kwargs.dockerfile}",
+                    "DOCKER_REPOSITORY=${repo}/${kwargs.dockerfile}"]
+    if (kwargs.fromDockerfile) {
+        environs << "FROM_DOCKER_REPOSITORY=${repo}/${kwargs.fromDockerfile}"
+    }
+    withEnv(environs) {
+        sh "docker/jenkins/${script}"
+    }
+}
+
+def pushDockerhub(from_repo, to_repo='') {
+    to_repo = to_repo ?: from_repo
+    withCredentials([[$class: 'StringBinding',
+                      credentialsId: 'DOCKER_PASSWORD',
+                      variable: 'DOCKER_PASSWORD']]) {
+        withEnv(['DOCKER_USERNAME=mozjenkins',
+                 "FROM_DOCKER_REPOSITORY=${from_repo}",
+                 "DOCKER_REPOSITORY=${to_repo}"]) {
+            retry(2) {
+                sh 'docker/jenkins/push2dockerhub.sh'
+            }
+        }
+    }
+}
+
+def pushPrivateReg(region) {
+    def registryPorts = [usw: '5001', euw: '5000']
+    retry(3) {
+        // TODO Fix DEIS_APPS before merge
+        // DEIS_APPS=bedrock-dev,bedrock-stage,bedrock-prod
+        withEnv(['FROM_DOCKER_REPOSITORY=mozorg/bedrock_l10n',
+                 "PRIVATE_REGISTRIES=localhost:${registryPorts[region]}",
+                 'DEIS_APPS=bedrock-jenkinsfile-test']) {
+            sh 'docker/jenkins/push2privateregistries.sh'
+        }
+    }
+}
+
+def integrationTestJob(propFileName, appName, region) {
+    def testsBaseDir = 'docker/jenkins/properties/integration_tests'
+    def testsFileExt = '.properties'
+    return {
+        node {
+            unstash 'scripts'
+            unstash 'tests'
+            def fullFilename = "${testsBaseDir}/${propFileName}${testsFileExt}"
+            def testScript = "docker/jenkins/run_integration_tests.sh ${fullFilename}".toString()
+            withCredentials([[$class: 'UsernamePasswordMultiBinding',
```

Comment:
I'd like to see a list of all of these, but I just found this out on the web somewhere when looking for `withCredentials` and `Jenkinsfile`. 🤷‍♂️